File: java-solutions/src/leetcode/P151ReverseWords.java
package leetcode;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class P151ReverseWords {

    public String reverseWords(String s) {
        List<String> toks = Arrays.stream(s.split(" "))
                .filter(t -> !t.isEmpty())
                .collect(Collectors.toList());
        Collections.reverse(toks);
        return String.join(" ", toks);
    }

    public static void main(String[] args) {
        P151ReverseWords sol = new P151ReverseWords();
        String s = new String("   hello   world   ");

        String res = sol.reverseWords(s);
        System.out.println(":" + res + ":");
    }
}

File: rust-solutions/leetcode/src/bin/p151-reverse-words.rs
/// # Problem 151 - reverse words in a string of words
/// https://leetcode.com/problems/reverse-words-in-a-string/

struct Solution;

impl Solution {
    pub fn reverse_words(s: String) -> String {
        let mut reversed_words: Vec<&str> = s.split_whitespace()
            .filter(|&t| !t.is_empty())
            .rev()
            .collect();
        reversed_words[..].join(" ")
    }
}

fn main() {
    let s = String::from("  hello  world  ");
    let mut r: Vec<&str> = s.split_whitespace()
        .filter(|&t| !t.is_empty())
        .rev()
        .collect();
    let joined = r.into_iter().fold(String::new(), |mut s, c| {
        s.push_str(c);
        s
    });
    dbg!(&joined);
}